File: sdk-java/src/test/java/io/superdurable/dex/integ/WorkflowUncompletedStateTimeoutWorkflow.java
/*
 * Portions of this file are derived from indeedeng/iwf-java-sdk.
 * Those portions are licensed under the Apache License, Version 2.0.
 * See LICENSES/Apache-2.0.txt and LEGACY_NOTICES.md.
 *
 * Modifications Copyright (c) 2026 Super Durable, Inc.
 *
 * Modifications are licensed under the Super Durable Source License 1.0.
 * Third-Party Materials remain under the Apache License, Version 2.0.
 * See LICENSE and LEGACY_NOTICES.md.
 */

package io.superdurable.dex.integ;

import io.superdurable.dex.Context;
import io.superdurable.dex.Flow;
import io.superdurable.dex.RetryPolicy;
import io.superdurable.dex.Step;
import io.superdurable.dex.StepDurability;
import io.superdurable.dex.StepList;
import io.superdurable.dex.StepDecision;
import io.superdurable.dex.StepOptions;

import java.time.Duration;

final class WorkflowUncompletedStateTimeoutWorkflow implements Flow<Integer> {
    private final WorkflowUncompletedStateTimeoutStep start =
            new WorkflowUncompletedStateTimeoutStep();

    @Override
    public StepList<Integer> getSteps() {
        return StepList.startStep(start);
    }
}

final class WorkflowUncompletedStateTimeoutStep implements Step<Integer> {
    @Override
    public Class<Integer> getInputType() {
        return Integer.class;
    }

    @Override
    public StepDecision execute(final Context context, final Integer input) {
        try {
            Thread.sleep(Duration.ofSeconds(2).toMillis());
        } catch (InterruptedException exception) {
            Thread.currentThread().interrupt();
            throw new IllegalStateException("timeout callback interrupted", exception);
        }
        return StepDecision.gracefulComplete(input);
    }

    @Override
    public StepOptions getStepOptions() {
        return StepOptions.newBuilder()
                .executeMethodTimeout(Duration.ofSeconds(1))
                .executeRetry(RetryPolicy.newBuilder().maximumAttempts(1).build())
                .executeDurability(StepDurability.SYNC)
                .build();
    }
}

File: sdk-rust/crates/dex-sdk/tests/integ/workflow_uncompleted_state_timeout_workflow.rs
// Portions of this file are derived from indeedeng/iwf-java-sdk.
// Those portions are licensed under the Apache License, Version 2.0.
// See LICENSES/Apache-2.0.txt and LEGACY_NOTICES.md.
//
// Modifications Copyright (c) 2026 Super Durable, Inc.
//
// Modifications are licensed under the Super Durable Source License 1.0.
// Third-Party Materials remain under the Apache License, Version 2.0.
// See LICENSE and LEGACY_NOTICES.md.

use std::time::Duration;

use dex_sdk::{
    Context, Flow, HandlerResult, RetryPolicy, Step, StepDecision, StepDurability, StepList,
    StepOptions,
};

pub(crate) struct WorkflowUncompletedStateTimeoutWorkflow {
    start: WorkflowUncompletedStateTimeoutStep,
}

impl WorkflowUncompletedStateTimeoutWorkflow {
    pub(crate) fn new() -> Self {
        Self {
            start: WorkflowUncompletedStateTimeoutStep,
        }
    }
}

impl Flow for WorkflowUncompletedStateTimeoutWorkflow {
    type StartInput = i32;

    fn steps(&self) -> StepList<'_, Self::StartInput> {
        StepList::start(&self.start)
    }
}

struct WorkflowUncompletedStateTimeoutStep;

impl Step for WorkflowUncompletedStateTimeoutStep {
    type Input = i32;

    fn execute(&self, _context: &mut Context, input: i32) -> HandlerResult<StepDecision> {
        std::thread::sleep(Duration::from_secs(2));
        Ok(StepDecision::graceful_complete(input))
    }

    fn options(&self) -> StepOptions<Self::Input> {
        StepOptions::new()
            .execute_method_timeout(Duration::from_secs(1))
            .execute_retry(RetryPolicy::new().maximum_attempts(1))
            .execute_durability(StepDurability::Sync)
    }
}
